### PR TITLE
Add Zstd/LZ4/(GZIP write) support for region files

### PR DESCRIPTION
**Pull request by egg82**
Using zstd-jni: https://github.com/luben/zstd-jni
There's a new option in PaperConfig for this which defaults to false.
Since the patch hooks Mojang's versioning system for new compression methods, it
should be internally consistent, easily-updatable, and reversible with a force-upgrade.

This will likely conflict with plugins and programs expecting region files to be compressed using the current standard.

### CODE REVIEW

**Comment by egg82** [comment]
LZ4 PR here: https://github.com/PaperMC/Paper/pull/2376
Note that Zstd is optimized for compression, where LZ4 is optimized for decompression.

**Comment by hugmanrique** [comment]
Related #416

File: Spigot-Server-Patches/0406-Add-Zstd-support-for-region-files.patch [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+     }
+ 
+@@ -487,6 +503,13 @@ public class RegionFile implements AutoCloseable {
+ 
+     void writeOversizedData(int x, int z, NBTTagCompound oversizedData) throws IOException {
+         File file = getOversizedFile(x, z);
++        // Paper start - zstd support
++        if (PaperConfig.useZstdCompression) {
++            try (DataOutputStream out = new DataOutputStream(new BufferedOutputStream(new com.github.luben.zstd.ZstdOutputStream(new java.io.FileOutputStream(file)), 32 * 1024))) {
++                NBTCompressedStreamTools.writeNBT(oversizedData, out);
```
Comment (egg82) [inline Spigot-Server-Patches/0406-Add-Zstd-support-for-region-files.patch]: Need a return statement here

**Comment by PaulBGD** [comment]
As a user/developer, I don't like this. This option generates worlds that look like regular Minecraft worlds, but can never be loaded in vanilla without a conversion program. I feel like a major change to the filetype should result in a different filetype, even if that means just changing the extension of the files.

**Comment by wizjany** [comment]
To the contrary, the file format specifies a compression type. The fact that the vanilla implementation doesn't recognize this new type  doesn't mean the file is not-to-spec. Given that this isn't meant to be a default, I don't think admins using this option will expect it to load in vanilla.

**Comment by VADemon** [comment]
@PaulBGD I wanted to complain that a standalone tool is missing, this **is** the downside. But you can repack the whole world using Paper again as egg82 said:

> reversible with a `--force-upgrade`

**Comment by astei** [comment]
Personally, I see no reason not to include this. If the user decides to enable zstandard or LZ4 compression, they must understand that their setup will be non-standard and tools that manipulate the world data directly will not work.

**Comment by egg82** [comment]
Stats for the different algorithms:
Zlib: https://docs.google.com/spreadsheets/d/1rf_srqWDvEfBDR9OmcFofkcuqdqV5ypCZb8ImQ4L1FE/edit?usp=sharing
LZ4: https://docs.google.com/spreadsheets/d/1iNcqo07ySM-r2-1DTBOmd5vl809u97AmacdxSOMiPCQ/edit?usp=sharing
Zstd: https://docs.google.com/spreadsheets/d/1mGqXu2fAgQXiFgiQdE830TJayZwWbyz2yqYT4T4KRfo/edit?usp=sharing

Test bench used: https://github.com/egg82/CompressionTests
Training/test data used: [https://drive.google.com/open?id=1egzZKkV305LX7jt2Xxe0Y-DfYkkozDa_](https://drive.google.com/open?id=1egzZKkV305LX7jt2Xxe0Y-DfYkkozDa_)

**Comment by ghost** [comment]
Thank you for making this default to false.

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch [hunk 2]
```diff
@@ -0,0 +1,1123 @@
...
+     }
++
++    public static int compressionMethod = 2;
++    private static void compressionMethod() {
++        compressionMethod = getInt("settings.compression-method", 2);
++        if (compressionMethod < 0 || compressionMethod > 127 || (compressionMethod > 2 && compressionMethod < 126)) {
++            fatal("compression-method must be 0, 1, 2, 126, or 127");
++        }
++
++        if (compressionMethod == 127) {
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: static final constants for these

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch [hunk 3]
```diff
@@ -0,0 +1,1123 @@
...
+ 
++import com.destroystokyo.paper.PaperConfig;
+ import com.destroystokyo.paper.exception.ServerInternalException;
+ import com.google.common.collect.Lists;
+ import java.io.BufferedInputStream;
+@@ -12,6 +13,7 @@ import java.io.File;
+ import java.io.IOException;
+ import java.io.RandomAccessFile;
+ import java.util.List;
++import java.util.zip.DataFormatException;
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: needs //paper comment or removal, same for the paperconfig import

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch [hunk 4]
```diff
@@ -0,0 +1,1123 @@
...
+-        } catch (IOException ioexception) {
++        } catch (IOException | DataFormatException | com.github.luben.zstd.ZstdException ioexception) { // Paper - new & optimized compression
+             ServerInternalException.reportInternalException(ioexception); // Paper
+             return null;
+         }
+@@ -294,6 +386,7 @@ public class RegionFile implements AutoCloseable {
+ 
+             // Paper - move this into writeChunkData
+         } catch (IOException ioexception) {
++            ioexception.printStackTrace();
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: missing //paper comment

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch [hunk 5]
```diff
@@ -0,0 +1,1123 @@
...
+     }
+ 
+@@ -487,18 +580,58 @@ public class RegionFile implements AutoCloseable {
+ 
+     void writeOversizedData(int x, int z, NBTTagCompound oversizedData) throws IOException {
+         File file = getOversizedFile(x, z);
+-        try (DataOutputStream out = new DataOutputStream(new BufferedOutputStream(new DeflaterOutputStream(new java.io.FileOutputStream(file), new java.util.zip.Deflater(java.util.zip.Deflater.BEST_COMPRESSION), 32 * 1024), 32 * 1024))) {
+-            NBTCompressedStreamTools.writeNBT(oversizedData, out);
++        // Paper start - new & optimized compression
++        if (PaperConfig.compressionMethod == 127) {
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: switch statement is designed for this (as well as below)

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch [hunk 6]
```diff
@@ -0,0 +1,1123 @@
...
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -388,4 +388,32 @@ public class PaperConfig {
+         maxBookPageSize = getInt("settings.book-size.page-max", maxBookPageSize);
+         maxBookTotalSizeMultiplier = getDouble("settings.book-size.total-multiplier", maxBookTotalSizeMultiplier);
+     }
++
++    public static int compressionMethod = 2;
++    private static void compressionMethod() {
++        compressionMethod = getInt("settings.compression-method", 2);
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: we should use strings in the config to clearly indicate what lib will be used

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch (hunk 2 above)
Comment (egg82) [reply Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: enum with a string and an int instead. Then we can just enumerate the enum to print out all valid possibilities on the fatal.

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch (hunk 3 above)
Comment (egg82) [reply Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: Totally didn't accidentally leave in imports. Nope.

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch (hunk 4 above)
Comment (egg82) [reply Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: Totally didn't accidentally leave in debug stuff. Nope.

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch (hunk 6 above)
Comment (egg82) [reply Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: See enum comment above

**Comment by egg82** [comment]
> Thank you for making this default to false.

There's no way anyone here would want the default to be something other than what Mojang uses. It's far too risky with external tools and plugins.

File: Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch [hunk 7]
```diff
@@ -0,0 +1,1194 @@
...
+index 000000000..fd7bd51b4
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/region/RegionCompressionType.java
+@@ -0,0 +1,43 @@
++package com.destroystokyo.paper.region;
++
++import java.util.Optional;
++
++public enum RegionCompressionType {
++    NOP(0, "nop"),
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Add-Zstd-LZ4-GZIP-write-support-for-region-files.patch]: noop or none

**Comment by LoneDev6** [comment]
Do you think it will conflict with https://www.spigotmc.org/resources/regionerator.12219/ ?

**Comment by egg82** [comment]
> Do you think it will conflict with https://www.spigotmc.org/resources/regionerator.12219/ ?

Almost certainly, yes. Plugins that read files directly (such as Regionerator, Dynmap, etc) won't be able to read the chunk. Since the patch uses the standard file format to determine what compression algorithm is used, some will (hopefully) fail silently. If not, contact the author. Plugins will have access to the Zstd dictionary in Paper, and the patch is MIT.

**Pull request closed without merging**